clients/privacy-center/components/RequestModal.tsx

```diff
@@ -0,0 +1,293 @@
+import React, { useEffect, useState } from 'react';
+import {
+  Modal,
+  ModalOverlay,
+  ModalContent,
+  ModalHeader,
+  ModalFooter,
+  ModalBody,
+  Text,
+  Button,
+  chakra,
+  Stack,
+  FormControl,
+  Input,
+  FormErrorMessage,
+} from '@fidesui/react';
+
+import { useFormik } from 'formik';
+
+import type { AlertState } from '../types/AlertState';
+
+import config from '../config/config.json';
+
+export const useRequestModal = () => {
+  const [isOpen, setIsOpen] = useState(false);
+  const [openAction, setOpenAction] = useState<string | null>(null);
+
+  const onOpen = (action: string) => {
+    setOpenAction(action);
+    setIsOpen(true);
+  };
+
+  const onClose = () => {
+    setIsOpen(false);
+    setOpenAction(null);
+  };
+
+  return { isOpen, onClose, onOpen, openAction };
+};
+
+const useRequestForm = ({
+  onClose,
+  action,
+  setAlert,
+}: {
+  onClose: () => void;
+  action: typeof config.actions[0] | null;
+  setAlert: (state: AlertState) => void;
+}) => {
+  const [isLoading, setIsLoading] = useState(false);
+  const formik = useFormik({
+    initialValues: {
+      name: '',
+      email: '',
+      phone: '',
+    },
+    onSubmit: async (values) => {
+      if (!action) {
+        // somehow we've reached a broken state, return
+        return;
+      }
+
+      setIsLoading(true);
+      const host =
+        process.env.NODE_ENV === 'development'
+          ? config.fidesops_host_development
+          : config.fidesops_host_production;
+      const body = [
+        {
+          identity: {
+            email: values.email,
+            phone_number: values.phone,
+            // enable this when name field is supported on the server
+            // name: values.name
+          },
+          policy_key: action.policy_key,
+        },
+      ];
+
+      const handleError = () => {
+        setAlert({
+          status: 'error',
+          description: 'Your request has failed. Please try again.',
+        });
+        onClose();
+      };
+
+      try {
+        const response = await fetch(`${host}/privacy-request`, {
+          method: 'POST',
+          headers: {
+            Accept: 'application/json',
+            'Content-Type': 'application/json',
+          },
+          body: JSON.stringify(body),
+        });
+
+        if (!response.ok) {
+          handleError();
+          return;
+        }
+
+        const data = await response.json();
+
+        if (data.succeeded.length) {
+          setAlert({
+            status: 'success',
+            description:
+              'Your request was successful, please await further instructions.',
+          });
+        } else {
+          handleError();
+        }
+      } catch (error) {
+        handleError();
+        return;
+      }
+
+      onClose();
+    },
+    validate: (values) => {
```

Comment:
What are your thoughts on invoking this function on form submit, rather than on unfocus of an input, as this results in weird looking scenarios like:

![Screenshot 2022-03-10 at 14 15 01](https://user-images.githubusercontent.com/1667340/157737885-894f7d26-cd16-420c-bdb4-df5289a942e2.png)

Reply:
Form validation on field blur is a fairly standard pattern, and in fact it's the standard one that Formik (the form library we're using) recommends. I suspect this only looks strange because you're deliberately testing the input – most people will enter something before moving on to the next field. :)

If you'd prefer to validate the form on submission – and keep the "Continue" button active initially instead of having it become active when the form is valid – I can implement that. It would require reworking some of the other logic as well, but shouldn't be a problem. That's also a fairly common pattern.